### ISSUE TITLE
pd.DataFrame.__deepcopy__ is broken

### ISSUE THREAD

**Issue by mortonjt** [pandas-dev/pandas#17406]
Looks like the pandas deepcopy is no longer fully recursive.  For example if we have nested lists as shown below, the most inner lists are actually exactly the same.  

```python
>>> x = [[1, 2, 3, 4]]
>>> y = [[2, 3, 4, 5]]
>>> import pandas as pd
>>> df = pd.DataFrame({'x':x, 'y':y})
>>> import copy
>>> df2 = copy.deepcopy(df)
>>> id(df2['x'][0])
4572099912
>>> id(df['x'][0])
4572099912
```

We first noticed this issue on [skbio](https://github.com/biocore/scikit-bio/pull/1530) when the copy unittests were failing.  This hasn't been a problem with the previous pandas release.  Looking at the commits in the previous release, this [one](https://github.com/pandas-dev/pandas/commit/0b4fdf988e3125f7c55aaf6e08a2dfa7d9e2e8a0) looks suspicious.

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.5.3.final.0
python-bits: 64
OS: Darwin
OS-release: 15.6.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.20.3
pytest: None
pip: 9.0.1
setuptools: 27.2.0
Cython: None
numpy: 1.13.1
scipy: 0.19.1
xarray: None
IPython: 6.1.0
sphinx: None
patsy: 0.4.1
dateutil: 2.6.1
pytz: 2017.2
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 2.0.2
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 0.999
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.9.6
s3fs: None
pandas_gbq: None
pandas_datareader: None

</details>

**Comment by chris-b1** [comment]
Yeah, you definitely have the right change, our `.copy(deep=True)`, which `__deepcopy__`  now calls isn't fully recursive
```
In [108]: df.copy?
Signature: df.copy(deep=True)
Docstring:
Make a copy of this objects data.

Parameters
----------
deep : boolean or string, default True
    Make a deep copy, including a copy of the data and the indices.
    With ``deep=False`` neither the indices or the data are copied.

    Note that when ``deep=True`` data is copied, actual python objects
    will not be copied recursively, only the reference to the object.
    This is in contrast to ``copy.deepcopy`` in the Standard Library,
    which recursively copies object data.
```

**Comment by jreback** [comment]
embedding mutable objects inside a. DataFrame is an antipattern

**Comment by ebolyen** [comment]
@jreback, definitely agree. We use pandas to manage metadata about objects that have positions. We also implement `__deepcopy__` and so our tests specifically set up this situation to ensure everything matches the spec. We don't have a lot of control over what users might put here, and it absolutely could be argued that they *shouldn't* put something that is mutable there, but `__deepcopy__` did work in the past for pandas.

Out of curiosity, wouldn't it be better to not implement the `__deepcopy__` protocol if it isn't supported?

**Comment by jreback** [comment]
> Out of curiosity, wouldn't it be better to not implement the __deepcopy__ protocol if it isn't supported?

it was already there; this just allows a method to actually work on it. The fact that you cannot deepcopy mutable objects is clearly indicated. I think numpy also does this.

Would take a patch to fix it though.

**Comment by asishm** [comment]
@jreback Just my view, but there are methods in pandas that return mutable objects, a notable one being `.str.split`. I would say that while having mutable objects at the start/end of a pipeline may be an antipattern, but it definitely shows up in intermediate steps. I can't think of any other use-cases right now, but a lot of NLP tools (word embeddings, tokenizers) etc. do return mutable objects as well.

**Comment by adrianog** [comment]
What's the latest on this? Is there an official workaround to perform a deepcopy [apart from](https://stackoverflow.com/a/55031463/5579463):
```import copy
df2 = df.apply(copy.deepcopy)```

**Comment by jreback** [comment]
the community is welcome to contribute a patch

**Comment by miladad8** [comment]
now that deep copy doesn't work, how can we  create an independent dataframe from an existing one so that modifcations in copied one  wouldn't affect the original dataframe?  really appreciate if you help .

**Comment by PlatonB** [comment]
> What's the latest on this? Is there an official workaround to perform a deepcopy [apart from](https://stackoverflow.com/a/55031463/5579463):
> 
> ```python
> import copy
> df2 = df.apply(copy.deepcopy)
> ```

@adrianog It doesn't work either:
```
>>> import copy
>>> import pandas as pd
>>> df = pd.DataFrame({'q': ['a', 'b', 'c'], 'w': [1, 2, [3, 4, 5]]})
>>> df
   q          w
0  a          1
1  b          2
2  c  [3, 4, 5]
>>> df_c = df.apply(copy.deepcopy)
>>> df_c
   q          w
0  a          1
1  b          2
2  c  [3, 4, 5]
>>> del df_c.loc[2, 'w'][0]
>>> df_c
   q       w
0  a       1
1  b       2
2  c  [4, 5]
>>> df
   q       w
0  a       1
1  b       2
2  c  [4, 5]
>>>
```
